halo/app/app.go

```diff
@@ -254,34 +259,46 @@ func dumpLastAppliedUpgradeInfo(ctx sdk.Context, keeper *upgradekeeper.Keeper) e
 	return nil
 }
 
-// isErrOldBinary returns the true if the error is due to the
+// isErrOldBinary returns the last applied upgrade and true if the error is due to the
 // upgrade module detecting the binary is too old.
-func isErrOldBinary(err error) bool {
+func isErrOldBinary(err error) (string, bool) {
 	if err == nil {
-		return false
+		return "", false
 	}
 
-	// Get the root cause of the error.
-	cause := err
-	for {
-		next := errors.Unwrap(cause)
-		if next == nil {
-			break
-		}
-		cause = next
+	cause := errors.Cause(err)
+
+	reg := regexp.MustCompile(`wrong app version \d+, upgrade handler is missing for (.+) upgrade plan`)
+	if !reg.MatchString(cause.Error()) {
+		return "", false
+	}
+
+	matches := reg.FindStringSubmatch(cause.Error())
+	if len(matches) != 2 {
+		return "", false
+	}
+
+	return matches[1], true
+}
+
+func isErrUpgradeNeeded(err error) (string, bool) {
+	if err == nil {
+		return "", false
+	}
+
+	cause := errors.Cause(err)
+
+	reg := regexp.MustCompile(`UPGRADE "(.+)" NEEDED .+`)
+	if !reg.MatchString(cause.Error()) {
+		return "", false
 	}
 
-	var d int
-	var s string
-	i, err := fmt.Fscanf(
```

Comment:
fmt.Scan is the poor mans regexp, it doesn't support spaces in strings for example